File: mockboard.py
import random
from generate_stock import StockProperties


def print_options_board(calls, strikes, puts):
    # Find the maximum width for each column
    call_width = 5
    strike_width = 7
    put_width = 5

    # Print the header
    print(
        f"{'calls':>{call_width}}  {'strikes':>{strike_width}}  {'puts':>{put_width}}"
    )
    # Print the rows
    for call, strike, put in zip(calls, strikes, puts):
        print(
            f"{call:>{call_width}.2f}  {strike:^{strike_width}}  {put:>{put_width}.2f}"
        )


class MockBoard:
    def __init__(self, stockinfo: StockProperties, stock=None, quote=None):
        self.stockinfo = stockinfo
        self.stock = stock
        if stock is None:
            self.stock = self.stockinfo.stock
        self.quote = quote

    def print_board(self):
        stock = self.stock
        rc = self.stockinfo.rc
        strikes = self.stockinfo.strikes
        calls, puts = self.stockinfo.get_options(stock)

        if self.quote is not None:
            print(self.quote)
        else:
            print(f"S = {stock:.2f}")
        print(f"r/c = {rc:.2f}")

        bw = f"{strikes[0]} B/W={puts[0] + rc:.2f}"
        ps = f"{strikes[1]}/{strikes[2]} PS={puts[2] - puts[1]:.2f}"
        strad = f"{strikes[2]} ∀={calls[2] + puts[2]:.2f}"
        cs = f"{strikes[2]}/{strikes[3]} CS={calls[2] - calls[3]:.2f}"
        pns = f"{strikes[4]} P+S={calls[4] - rc:.2f}"

        print(" " * 19 + f"{strikes[0]:<6}{bw:19}")
        print(" " * 19 + f"{strikes[1]:<6}{ps:19}")
        print(f"{cs:19}{strikes[2]:<6}{strad:19}")
        print(" " * 19 + f"{strikes[3]:<6}")
        print(f"{pns:19}{strikes[4]:<6}")

        print("Press enter to see answers...")
        input()
        print_options_board(calls, strikes, puts)

    def make_markets(self):
        stock = self.stock
        rc = self.stockinfo.rc
        strikes = self.stockinfo.strikes
        calls, puts = self.stockinfo.get_options(stock)
        print("Press enter to make markets...")
        while True:
            action = input()
            if action == "q":
                break
            k = random.choice(strikes)
            i = random.randint(0, 3)
            fair = 0
            if random.random() < 0.5:
                fair = abs(stock - k + rc)
                print(f"Make a market in {k} combos", end="\r")
            else:
                ins = random.choice(["RR", "CS", "PS"])
                print(f"Make a market in {strikes[i]}/{strikes[i + 1]} {ins}", end="\r")
                if ins == "RR":
                    fair = abs(puts[i] - calls[i + 1])
                elif ins == "CS":
                    fair = calls[i] - calls[i + 1]
                elif ins == "PS":
                    fair = puts[i + 1] - puts[i]

            answer = input()
            print(f"fair price = {fair:.2f}")


if __name__ == "__main__":
    stockinfo = StockProperties()
    mockboard = MockBoard(stockinfo)
    mockboard.print_board()
    mockboard.make_markets()

File: src/mockboard.js
class MockBoard {
	constructor(stockinfo, stock = null, quote = null) {
		this.stockinfo = stockinfo;
		this.stock = stock || this.stockinfo.stock;
		this.quote = quote;
	}

	print_board() {
		const stock = this.stock;
		const rc = this.stockinfo.rc;
		const strikes = this.stockinfo.strikes;
		const [calls, puts] = this.stockinfo.get_options(stock);
		this.calls = calls;
		this.puts = puts;

		const output1 = document.getElementById("output1");
		const tbody = document.getElementById('optionsTable');
		tbody.innerHTML = '';
		let bw = `${strikes[0]} B/W=${(puts[0] + rc).toFixed(2)}`;
		let ps = `${strikes[1]}/${strikes[2]} PS=${(puts[2] - puts[1]).toFixed(2)}`;
		let strad = `${strikes[2]} ∀=${(calls[2] + puts[2]).toFixed(2)}`;
		let cs = `${strikes[2]}/${strikes[3]} CS=${(calls[2] - calls[3]).toFixed(2)}`;
		let pns = `${strikes[4]} P+S=${(calls[4] - rc).toFixed(2)}`;
		const rows = [["","","","",strikes[0],"","","",bw],
					 ["","","","",strikes[1],"","","",ps],
					 [cs,"","","",strikes[2],"","","",strad],
					 ["","","","",strikes[3],"","","",""],
			         [pns,"","","",strikes[4],"","","",""]];

		rows.forEach(values => {
			const row = document.createElement("tr");
			values.forEach( value => {
				const cell = document.createElement("td");
				cell.textContent = value; // Set cell content
    		    row.appendChild(cell);	
			});
			tbody.appendChild(row);
		});
	}

	show_theos() {
		const tbody = document.getElementById('optionsTable');
		const rows = tbody.rows;
		for (let i = 0; i < 5; i++) {
			const row = rows[i];
			const callcell = row.cells[2];
			const putcell = row.cells[6];
			callcell.textContent = this.calls[i].toFixed(2);
			putcell.textContent = this.puts[i].toFixed(2);
		}
	}

	hide_theos() {
		const tbody = document.getElementById('optionsTable');
		const rows = tbody.rows;
		for (let i = 0; i < 5; i++) {
			const row = rows[i];
			const callcell = row.cells[2];
			const putcell = row.cells[6];
			callcell.textContent = "";
			putcell.textContent = "";
		}
	}

	make_markets() {
		const stock = this.stock;
		const rc = this.stockinfo.rc;
		const strikes = this.stockinfo.strikes;
		const [calls, puts] = this.stockinfo.get_options(stock);
		const output1 = document.getElementById("output1");
		const inputBox = document.getElementById("inputBox");

		output1.innerHTML = "Press enter to make markets...";

		inputBox.addEventListener("keydown", (e) => {
			if (e.key === "Enter") {
				let action = inputBox.value;
				if (action === "q") return;

				let k = strikes[Math.floor(Math.random() * strikes.length)];
				let i = Math.floor(Math.random() * 4);
				let fair = 0;
				let ins = "";

				if (Math.random() < 0.5) {
					fair = Math.abs(stock - k + rc);
					output1.innerHTML = `Make a market in ${k} combos`;
				} else {
					ins = ["RR", "CS", "PS"][Math.floor(Math.random() * 3)];
					output1.innerHTML = `Make a market in ${strikes[i]}/${strikes[i + 1]} ${ins}`;
					if (ins === "RR") {
						fair = Math.abs(puts[i] - calls[i + 1]);
					} else if (ins === "CS") {
						fair = calls[i] - calls[i + 1];
					} else if (ins === "PS") {
						fair = puts[i + 1] - puts[i];
					}
				}
				output1.innerHTML += `<br>fair price = ${fair.toFixed(2)}`;
			}
		});
	}
}

